cmake/GenerateVersion.cmake

```diff
@@ -1,33 +1,33 @@
-if(NVIM_VERSION_MEDIUM)
-  message(STATUS "USING NVIM_VERSION_MEDIUM: ${NVIM_VERSION_MEDIUM}")
-  return()
-endif()
-
-set(NVIM_VERSION_MEDIUM
+set(NVIM_VERSION
     "v${NVIM_VERSION_MAJOR}.${NVIM_VERSION_MINOR}.${NVIM_VERSION_PATCH}${NVIM_VERSION_PRERELEASE}")
 
 execute_process(
   COMMAND git describe --first-parent --dirty --always
   OUTPUT_VARIABLE GIT_TAG
+  OUTPUT_STRIP_TRAILING_WHITESPACE
   RESULT_VARIABLE RES)
 
 if(RES AND NOT RES EQUAL 0)
-  message(STATUS "Using NVIM_VERSION_MEDIUM: ${NVIM_VERSION_MEDIUM}")
-  file(WRITE "${OUTPUT}" "${NVIM_VERSION_STRING}")
+  message(STATUS "Using NVIM_VERSION: ${NVIM_VERSION}")
+  file(WRITE "${OUTPUT}" "")
   return()
 endif()
 
-string(STRIP "${GIT_TAG}" GIT_TAG)
-string(REGEX REPLACE "^v[0-9]+.[0-9]+.[0-9]+-" "" NVIM_VERSION_GIT "${GIT_TAG}")
-set(NVIM_VERSION_MEDIUM "${NVIM_VERSION_MEDIUM}-${NVIM_VERSION_GIT}")
-set(NVIM_VERSION_STRING "#define NVIM_VERSION_MEDIUM \"${NVIM_VERSION_MEDIUM}\"\n")
+# `git describe` annotates the most recent tagged release; for pre-release
+# builds we append that to the dev version.
+if(NVIM_VERSION_PRERELEASE)
+  string(REGEX REPLACE "^v[0-9]+.[0-9]+.[0-9]+-" "" NVIM_VERSION_GIT "${GIT_TAG}")
+  set(NVIM_VERSION "${NVIM_VERSION}+${NVIM_VERSION_GIT}")
```

Comment:
is this changing the hyphen to a `+` ? The choice of hyphen was intentional based on https://semver.org/ . Anything after a `+` is treated as arbitrary and not part of a version calculation, whereas a hyphen _is_ part of the version calculation.

Reply:
Yes, because the old version generation had a plus as well. Shall I change it back to minus?

Reply:
This PR will make the version look like

```
v0.9.0-dev+67-g625ba79be
```

, which is how it used to look with `scripts/update_version_stamp.lua` while the one in master currently looks like

```
v0.9.0-dev-67-g625ba79be
```

I didn't mention this in the commit message because I thought it was purely a stylistic choice. I will update the commit message.

Reply:
Wouldn't
```
v0.9.0-dev-67+g625ba79be
```
be the correct format then (`67` is a strictly increasing commit counter, so larger == later, but `g625ba79be` is a commit hash that shouldn't be compared for ordering purposes)?

Reply:
Strictly speaking, `dev` is redundant with the commit counter, but dropping it would needlessly break tooling (ours and downstream), I'm sure.

Reply:
> Strictly speaking, dev is redundant with the commit counter, but dropping it would needlessly break tooling (ours and downstream), I'm sure.

I think this is to let package maintainers without a `.git` folder/history know that their version isn't a release version. A bit academic example, but that's the only thing I can think of.

Reply:
> This PR will make the version look like `v0.9.0-dev+67-g625ba79be`
> , which is how it used to look

Ok then I mis-remembered. Whatever we had chosen years ago was the intentional, correct form.

Reply:
> I think this is to let package maintainers without a .git folder/history know that their version isn't a release version.

`-dev` is important for when users, for whatever reason, didn't get a build with a git version, so that we know they aren't on an actual release.

Reply:
Holy cow, the neovim version has been on a journey.

So in [neovim/neovim@`b4b4536` (#3839)](https://github.com/neovim/neovim/pull/3839/commits/b4b45363396ed717230a78f8a4968e2dd12c014a) a dev version was `0.1.2-dev` for non-git builds and `v0.1.2-176-g9c3c2b5` for git builds. This according to the commit message, haven't tried it out myself.

Then in https://github.com/neovim/neovim/commit/2b594b9c81fea02598e0f6b6dff7972677f2c150 this was changed so that `-dev` was always added to a git build. The `-` was also changed to a `+`. The reason for these changes aren't explained.

Then in https://github.com/neovim/neovim/pull/19289 the `+` was changed to a `-`, likely due to a mistake.

What is the neovim version even supposed to be? I did a quick search on past discussions but couldn't really find anything conclusive other than semver specification pedantry for a lack of a better term.

Maybe we should write this down somewhere for future reference (`MAINTAIN.md` maybe?)

Reply:
Let's put the discussion about the correct versioning on the shelf for another time and just focus on release getting a real version :)